src/user.cc

```diff
@@ -1064,23 +1064,29 @@ void User::loadUserTimeEntryFromJSON(
         return;
     }
 
-    TimeEntry *model = related.TimeEntryByID(id);
+    TimeEntry* model;
+    {
+        Poco::Mutex::ScopedLock lock(loadTimeEntries_m_);
+        model = related.TimeEntryByID(id);
 
-    if (!model) {
-        model = related.TimeEntryByGUID(data["guid"].asString());
-    }
+        if (!model) {
+            model = related.TimeEntryByGUID(data["guid"].asString());
+        }
 
-    if (!data["server_deleted_at"].asString().empty()) {
-        if (model) {
-            model->MarkAsDeletedOnServer();
+        if (!data["server_deleted_at"].asString().empty()) {
+            if (model) {
+                model->MarkAsDeletedOnServer();
+            }
+            return;
         }
-        return;
-    }
 
-    if (!model) {
-        model = new TimeEntry();
-        related.pushBackTimeEntry(model);
+        if (!model) {
+            model = new TimeEntry();
+            model->SetID(id);
```

Comment:
I've added this one line. `ID` was previously set in `LoadFromJSON()`, but now we want to make sure all entries in `related` have correct IDs before we get out of scope of the lock, so that `related.TimeEntryByID()` returns correct results.

Reply:
The code it self is straightforward and this context is locked properly. 

However, I'm not able to reproduce the Constrain issues itself, so I'm not guarantee that this fix could completely fix it or now.